br/pkg/lightning/backend/external/onefile_writer.go

```diff
@@ -0,0 +1,240 @@
+	return err
+}
+
+// WriteRow implements ingest.Writer.
+func (w *OneFileWriter) WriteRow(ctx context.Context, idxKey, idxVal []byte, handle tidbkv.Handle) error {
+	keyAdapter := w.keyAdapter
+
+	var rowID []byte
+	if handle != nil {
+		rowID = handle.Encoded()
+	}
+	encodedKeyLen := keyAdapter.EncodedLen(idxKey, rowID)
+	length := encodedKeyLen + len(idxVal) + lengthBytes*2
+	blockIdx, dataBuf, off, allocated := w.kvBuffer.Alloc(length)
+	if !allocated {
+		if err := w.flushKVs(ctx, false); err != nil {
+			return err
+		}
+		blockIdx, dataBuf, off, allocated = w.kvBuffer.Alloc(length)
+		// we now don't support KV larger than blockSize
+		if !allocated {
+			return errors.Errorf("failed to allocate kv buffer: %d", length)
+		}
+	}
+	binary.BigEndian.AppendUint64(dataBuf[:0], uint64(encodedKeyLen))
+	keyAdapter.Encode(dataBuf[lengthBytes:lengthBytes:lengthBytes+encodedKeyLen], idxKey, rowID)
+	binary.BigEndian.AppendUint64(dataBuf[lengthBytes+encodedKeyLen:lengthBytes+encodedKeyLen], uint64(len(idxVal)))
+	copy(dataBuf[lengthBytes*2+encodedKeyLen:], idxVal)
+
+	w.kvLocations = append(w.kvLocations, kvLocation{
+		blockIdx: blockIdx,
+		offset:   off,
+		length:   int32(length)},
+	)
+	w.kvSize += int64(encodedKeyLen + len(idxVal))
+	w.batchSize += uint64(length)
+	return nil
+}
+
+// LockForWrite implements ingest.Writer.
+// Since flushKVs is thread-safe in external storage writer,
+// this is implemented as noop.
+func (w *OneFileWriter) LockForWrite() func() {
+	return func() {}
+}
+
+// Close closes the writer.
+func (w *OneFileWriter) Close(ctx context.Context) error {
+	if w.closed {
+		return errors.Errorf("writer %s has been closed", w.writerID)
+	}
+	w.closed = true
+	defer w.kvBuffer.destroy()
+	err := w.flushKVs(ctx, true)
+	if err != nil {
+		return err
+	}
+	w.logger.Info("close one file writer",
+		zap.String("writerID", w.writerID),
+		zap.Int("kv-cnt-cap", cap(w.kvLocations)),
+		zap.String("minKey", hex.EncodeToString(w.minKey)),
+		zap.String("maxKey", hex.EncodeToString(w.maxKey)))
+
+	w.kvLocations = nil
+
+	w.onClose(&WriterSummary{
+		WriterID:           w.writerID,
+		Seq:                0,
+		Min:                w.minKey,
+		Max:                w.maxKey,
+		TotalSize:          w.totalSize,
+		MultipleFilesStats: []MultipleFilesStat{w.multiFileStat},
+	})
+	return nil
+}
+
+func (w *OneFileWriter) recordMinMax(newMin, newMax tidbkv.Key, size uint64) {
+	if len(w.minKey) == 0 || newMin.Cmp(w.minKey) < 0 {
+		w.minKey = newMin.Clone()
+	}
+	if len(w.maxKey) == 0 || newMax.Cmp(w.maxKey) > 0 {
+		w.maxKey = newMax.Clone()
+	}
+	w.totalSize += size
+}
+
+func (w *OneFileWriter) flushKVs(ctx context.Context, fromClose bool) (err error) {
+	if len(w.kvLocations) == 0 {
+		return nil
+	}
+	
+	for _, pair := range w.kvLocations {
+		err = w.kvStore.addEncodedData(w.getEncodedKVData(pair))
+		if err != nil {
+			return err
+		}
+	}
+
+	w.kvStore.Close()
+	encodedStat := w.rc.encode()
```

Comment:
Maybe we can only write the statistic after all data is written for the data file, in other words, move it to `fromClose == true` branch

Reply:
stat file may very large

Reply:
stat file is quite small for a 256m data file in our test